### PR TITLE
geo objects textures

### PR DESCRIPTION
**Pull request by pavletto**

### CODE REVIEW

File: blender.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    for obj in collection.all_objects:
+        if obj.type == "MESH":
+            triangulate_object(obj)
+            matrix = obj.matrix_world
+            maxIndex = max(indices) + 1 if len(indices) > 0 else 0
+
+            for vert in obj.data.vertices:
+                gCoords = matrix @ vert.co
+                verts += list(gCoords.xyz)
+                normals += list(vert.normal)
```
Comment (Zemledelec) [inline blender.py:70]: Just an idea, add smooth and fat normal options (if blender doesn't provide it in the api). If "smooth" - use blender indexes, if "flat" use custom indexes, but before parsing all triangles individually from the current indexes set.

File: src/og/entity/GeoObject.js [hunk 2]
```diff
@@ -56,6 +62,8 @@ class GeoObject {
         this._handlerIndex = -1;
         this._vertices = options.vertices;
         this._normals = options.normals;
+        this._texCoords = options.texCoords || [];
+        // this._texCoordsIndices = options.texCoordsIndices || [];
```
Comment (Zemledelec) [inline src/og/entity/GeoObject.js]: Useless(confusing) commentary

File: src/og/entity/GeoObjectHandler.js [hunk 3]
```diff
@@ -50,6 +53,8 @@ class GeoObjectHandler {
         this._indicesArr = [[]];
         this._pickingColorArr = [[]];
         this._vehicleVisibleArr = [[]];
+        this._texCoordArr = [[]];
+        this._tcoordIndicesArr = [[]];
```
Comment (Zemledelec) [inline src/og/entity/GeoObjectHandler.js]: Unused property.

File: src/og/entity/GeoObjectHandler.js [hunk 4]
```diff
@@ -62,6 +67,8 @@ class GeoObjectHandler {
         this._indicesBuffer = [];
         this._pickingColorBuffer = [];
         this._vehicleVisibleBuffer = [];
+        this._texCoordBuffer = [];
+        // this._tcoordIndicesBuffer = [];
```
Comment (Zemledelec) [inline src/og/entity/GeoObjectHandler.js]: Cleanup

File: src/og/entity/GeoObjectHandler.js [hunk 5]
```diff
@@ -190,6 +219,20 @@ class GeoObjectHandler {
         }
     }
 
+    async setTexture(geoObject) {
```
Comment (Zemledelec) [inline src/og/entity/GeoObjectHandler.js:220]: I'd name it more sophisticated, loadGeoObjectTexture(geoObject), but it's ok

File: src/og/entity/GeoObjectHandler.js [hunk 6]
```diff
@@ -370,6 +424,8 @@ class GeoObjectHandler {
             gl.bindBuffer(gl.ARRAY_BUFFER, this._vehicleVisibleBuffer[ti]);
             gl.vertexAttribPointer(a.aDispose, this._vehicleVisibleBuffer[ti].itemSize, gl.FLOAT, false, 0, 0);
 
+            gl.uniform1f(u.uUseTexture, this._textures[ti] ? 1 : 0);
```
Comment (Zemledelec) [inline src/og/entity/GeoObjectHandler.js:422]: Remove uUseTexture

File: src/og/shaders/geoObject.js [hunk 7]
```diff
@@ -57,13 +65,18 @@ export const geo_object = () =>
...
             varying float vDispose;
+            varying float vUseTexture;
+            varying vec2 vTexCoords;
             
             const float RADIANS = 3.141592653589793 / 180.0;
-
+           
             void main(void) {
-            
+                vUseTexture = uUseTexture;
```
Comment (Zemledelec) [inline src/og/shaders/geoObject.js:74]: remove

File: src/og/utils/shared.js [hunk 8]
```diff
@@ -940,4 +940,21 @@ export function cloneArray(items) {
...
     return obj === void 0;
+}
+
+/**
+ * Promise for load images
+ * @function
+ * @param {string} url - link to image.
+ * @returns {Promise<Image>} Returns promise.
+ */
+export async function loadImage(url) {
```
Comment (Zemledelec) [inline src/og/utils/shared.js:951]: Btw, we have loader(Loader)for images and binary data etc, so might be better change it in future.

**Review by Zemledelec — CHANGES_REQUESTED** [review]
Cool! Remove useless uniforms.

**Review by Zemledelec — APPROVED** [review]

**Comment by Zemledelec** [comment]
@pavletto merging?

File: src/og/entity/GeoObjectHandler.js (hunk 6 above)
Comment (pavletto) [reply src/og/entity/GeoObjectHandler.js:422]: this uniform used for mixing colors and textures

**Pull request merged**